### ISSUE TITLE
Paypal Express: checkout.js execute call, permission denied, sometimes works 2nd time.

### ISSUE THREAD

**Issue by shicks3** [paypal/paypal-checkout#494]
I'm using Express Checkout, checkout.js v4.0.133 on a simple webpage, collecting membership dues.  Sometimes it works.  Most times it will only work with the second try after an error.   Issue occurs across browsers.

The form is on a wordpress page, simple form with some javascript to calculate feed.  It populates these variables which are used below: amountdue,	handling_fee, total, registrationId.

On first click of the paypal checkout, the Paypal dialog opens, lets me log in and shows the next screen with the $ amount.

![](https://i.stack.imgur.com/ccCZj.png)

I click on Pay Now and it falls into the error function with the following in the `console:`
```
  api.js:104 POST https://www.sandbox.paypal.com/webapps/hermes/api/payment/PAY-31P366013H791664VLHNNBMI/execute 401 (Unauthorized)
    index.js:572 api_retry_401 {method: "post", uri: "/webapps/hermes/api/payment/PAY-31P366013H791664VLHNNBMI/execute", timestamp: 1507512555418, windowID: "a2e02cce0b", pageID: "0a87b927a2"}
    api.js:104 POST https://www.sandbox.paypal.com/webapps/hermes/api/payment/PAY-31P366013H791664VLHNNBMI/execute 400 (Bad Request)
```
The first execute returns:

```
	{ack: "permission_denied", message: "Permission Denied"}
	    ack:"permission_denied"
	    message:"Permission Denied"
```

The second returns:

```
	{ack: "contingency", contingency: "DUPLICATE_TRANSACTION", meta: {calc: "febe5609a9e63",…},…}
	    ack:"contingency"
	    contingency:"DUPLICATE_TRANSACTION"
	    meta:{calc: "febe5609a9e63",…}
	        calc:"febe5609a9e63"
	        rlog:"a really long string of characters"
	    server:"a really long string of characters
```

Then, back on the web page, (no refresh), I click on the Paypal Checkout button again, the paypal dialog shown above appears, already having logged me in with the same username/password.   Click on Pay No and most of the time, success.  Sometimes another 400 error.

This happens in sandbox and production.  1 out of 6 times it will go through the 2nd time in production.  1 and 10 times it will go through the first time.  

This happens in either sandbox or production.  When it does succeed, payments all show up in the correct places on the buyer's and seller's paypal.  So I'm pretty confident the keys are correct.

I've been working on this for a couple of days without success.  Any help appreciated!!!

Below is my code, I removed some of the extraneous

```
	// sample data
	amountdue = 25.00;
	handling_fee = 1.00;
	total = 26.00;
	registrationId = 1501;

	paypal.Button.render({

		style: { size: 'medium' },
		env: 'sandbox', // sandbox | production
		client: {
			sandbox:    'AaCjaTydQm5tXqu5Vxp2IlhqafpsVxMfCC4h2Nu2YjbTE7dgiLIhCFL2oH01f8GNbxKfottwI0FfRa7-',
			production: 'xxx'
		},

		// Show the buyer a 'Pay Now' button in the checkout flow
		commit: true,

		// payment() is called when the button is clicked
		payment: function(data, actions) {
			submitFormRegistrationData();      // ajax call to store the forms data.
		
			// Make a call to the REST api to create the payment
			return actions.payment.create({
				payment: {
					transactions: [
						{
							amount: { 
								total: total, 
								currency: 'USD',
								details: {
									subtotal: amountdue,
									handling_fee: handling_fee
								}
							},
							invoice_number: registrationId
						}
					]
				},
				experience: {
					input_fields: {
						no_shipping: 1
					}
				}
			});
		},

		// onAuthorize() is called when the buyer approves the payment
		onAuthorize: function(data, actions) {
			// Make a call to the REST api to execute the payment
			return actions.payment.execute().then(function() {
				var d = { id : registrationId,	paymentId : data.paymentID };
				post_update(d,data);   // post the success to the database
				
				var completedText = 'Your payment was successful!</br></br>Your payment number is</br> '+data.paymentID+'.';
				showModal('Payment Completed',completedText,true,pp_modal_ok);
				
			}, function(reason) {							// 2nd function for "then", rejection
				showModal('Payment Error',"Paypal was not able to finish processing the payment.</br>You have not been charged.</br>This is usually a problem on Paypal's end.</br>Please try again by closing this window and clicking on the Paypal button again.");
			});
		},
		onCancel: function(data,actions) {
			showModal('Cancelled', 'You cancelled your payment.</br>You may try again or use the print/mail options.',false);
		},
		 onError: function(err) {
			showModal('Error','An error occurred which surprised us too!  Kindly let us know if you are seeing this.');
		}
	}, '#paypal-button-container');

	function submitFormRegistrationData() {                             // this always works.
		var data = $("#registrationForm").serializeArray();
		if ( $('#pastMember').is(':checked') ) {
			data.push({name:'pastMember', value:'1'});
		}
		//console.log(data);
		data.push({name:'op', value:'wr'});
		var p = $.post( "/register/api/register.php",$.param(data), 
			function(data) {						/* success function */
					res = JSON.parse(data);
					registrationId = res.id;
				})
			.fail(function(data) {
				showModal('Apologies, We Goofed','Something went wrong.  Thought we tested it all!</br>If you have a moment, please email websupport@naswc.org and let us know about it.</br></br>Thank you!');
			})
			.done(function() {})
			.always(function() {});
	}
	
``` 
To reproduce, I test with the page: https://www.naswc.org/membership-test (currently set on sandbox)

I appreciate any help.  I've searched and not found a related question so I posted.  Thank you.

**Comment by bluepnume** [comment]
Is there any possibility `invoice_number` / `registrationId` is non-unique? As in, the same `invoice_number` value is passed multiple times?

**Comment by shicks3** [comment]
I have a php log that records all api calls to /register/api/register.php from submitFormRegistrationData() of the javascript.  A new registrationId is returned each time the paypal button is clicked, which calls submitFormRegistrationData()..  

When I get the second error (401)  as it retries, that is definitely the same registrationId and that 2nd time I will see the duplicate error. See above the two responses from the first and retried 2nd call.

The problem is that first 401 error.  Trying to figure out why it always comes back with "permission denied".   

Could it be the way (or where) I call submitFormRegistrationData()?  It does an ajax call but I don't know why that would mess with paypal's payment function.

Thank you for responding

**Comment by bluepnume** [comment]
Any chance you could grab me the `paypal-debug-id` header value from the 401 response?

**Comment by shicks3** [comment]
Here is the entire header in case it helps:
```

Cache-Control:no-cache
Connection:keep-alive, Keep-Alive
Content-Encoding:gzip
content-security-policy:default-src 'self' https://*.paypal.com https://*.paypalobjects.com 'unsafe-inline' 'unsafe-eval'; connect-src https://*.paypal.com https://nexus.ensighten.com http://connect.facebook.net https://connect.facebook.net https://www.facebook.com https://m.facebook.com https://*.paypalobjects.com; frame-src https://*.paypal.com https://*.paypalobjects.com https://*.cardinalcommerce.com https://staticxx.facebook.com http://staticxx.facebook.com https://www.facebook.com https://m.facebook.com; script-src https://*.paypal.com https://ak1s.abmr.net https://akamai.mathtag.com https://*.akamaihd.net https://*.paypalobjects.com 'unsafe-inline' 'unsafe-eval'; img-src https: data:; block-all-mixed-content
Content-Type:application/json; charset=utf-8
Date:Tue, 10 Oct 2017 03:05:37 GMT
ETag:W/"39-58512yuDbE095FT2BqDRu6KAM+s"
HTTP_X_PP_AZ_LOCATOR:sandbox.slc
Keep-Alive:timeout=5, max=97
Paypal-Debug-Id:c7f9bbea0613
Paypal-Debug-Id:c7f9bbea0613
Server:Apache
Set-Cookie:X-PP-SILOVER=; Expires=Thu, 01 Jan 1970 00:00:01 GMT
Set-Cookie:x-csrf-jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ0b2tlbiI6ImZMRUk5NDRzdzQ1cmgwcXd5WXlXQk9WcWR1WVB2SVNmME9xVzJwenRiQS0zbnZWZFRxZkxUdWRyeURRNExmSzhSQUNndC1GQUJCcTF4Y1MwM2dqLTJRRE1MSC1YcmlMR1JNZVZBZDR3dlMzOVdpX21jOVZEZ2RLd2Q5SHUxWHhUMXNZSmRXM1hZRlJwenN4dDNhYmQ2MGxBUGgzR0dtampTQjl6OEtUWXpKd3A3X3ctblBlVkpOLUxqdFYwRllBZ2M4S3hWTE9CN0FyS3JxTjRoX0JZeGFEMVpDbSIsImlhdCI6MTUwNzYwNDczNywiZXhwIjoxNTA3NjA4MzM3fQ.upj35CHcs12EnNgHMl7RqKQBTkCLrFd6q9K9E_NTNqU; Domain=.paypal.com; Path=/; Expires=Tue, 17 Oct 2017 03:05:37 GMT; HttpOnly; Secure
Set-Cookie:jIHSLr4M5Yy4U-3CrkGq73ovwuW=S23AAGneRnztuj8Uz9wwSjB0HMUu-fT8FJ0Elqf6F9jbd1-PcS-lSuKbuEeWIrXkZg7M8WI74AfyYahq78uhoqbauYGsfDSKg; Domain=.paypal.com; Path=/; HttpOnly; Secure
Set-Cookie:X-PP-SILOVER=name%3DSANDBOX3.WEB.1%26silo_version%3D1880%26app%3Dhermesnodeweb%26TIME%3D20569177%26HTTP_X_PP_AZ_LOCATOR%3Dsandbox.slc; Expires=Tue, 10 Oct 2017 03:35:37 GMT; domain=.paypal.com; path=/; Secure; HttpOnly
Set-Cookie:tsrce=hermesnodeweb; Domain=.paypal.com; Path=/; Expires=Wed, 11 Oct 2017 03:05:37 GMT; HttpOnly; Secure
Set-Cookie:tsrce=hermesnodeweb; Domain=.paypal.com; Path=/; Expires=Wed, 11 Oct 2017 03:05:37 GMT; HttpOnly; Secure
Strict-Transport-Security:max-age=63072000
Transfer-Encoding:chunked
Vary:Accept-Encoding
X-Content-Type-Options:nosniff
x-csrf-jwt:eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ0b2tlbiI6IlA0eVlMeDFiVWRsSWtZdHlLY2NrUm9iUFpLZWMySzRGUExQTVJGb29aNGZiTWVBc1pSem1uZ1VTMk1NVnA0UXFOTWhnZG1sX19jX2FLMFdfUVJfTUlUM09TTjFuSnhwRU9Qc2pRYTExbkhOQ1BVemR3d1RnVzA3RFFfNTBxUkU4bmc1V09La1cteklldWVyVGpHV3B0cjJvUDJIWm1lRXpwYnJTdzdlRGFWTmNhTDRIaWNEUFdJRmFBQ0FJNVhPNE1qRGoybXloRlViMFJpTWlkRFRTaV9Ca0JqMCIsImlhdCI6MTUwNzYwNDczNywiZXhwIjoxNTA3NjA4MzM3fQ.VmV5G9UlK9_xFuABjZns9lmCIkI4lbDvqgKAnGJ309Q
x-csrf-jwt-hash:dc67de834fd1711ce5744d3f17ae1029abc5ca154039d788475e27a2ba74767f
X-Powered-By:Express
X-Recruiting:If you are reading this, maybe you should be working at PayPal instead! Check out https://www.paypal.com/us/webapps/mpp/paypal-jobs
x-xss-protection:1; mode=block
```

**Comment by bluepnume** [comment]
Hmm, strange, the log isn't coming through for that id. Could you possibly grab me a debug-id for production?

Appreciate the help on this -- not having much luck reproducing this in my tests.

**Comment by shicks3** [comment]
I appreciate YOUR help!!  :-)    Here is all the headers for production 401 error:
```
General
Request URL:https://www.paypal.com/webapps/hermes/api/payment/PAY-9KX0113746846031DLHOEQHI/execute
Request Method:POST
Status Code:401 
Remote Address:23.46.218.174:443
Referrer Policy:no-referrer-when-downgrade

Response Headers
paypal-debug-id:fc490805bf87
paypal-debug-id:fc490805bf87

Request Payload
{data: {payer_id: "RVQG8RN2DZN76"}, meta: {}}
    data:{payer_id: "RVQG8RN2DZN76"}
    meta:{}
```
And the response:
`{"ack":"permission_denied","message":"Permission Denied"}`

I have all the rest of the header, wasn't sure what to share on here since is production.

**Comment by shicks3** [comment]
Hi Daniel,
I got it to work consistently by commenting out the invoice_number from the transaction.  I really require it so if you have any suggestions as to how to include it, would be very much appreciated.  Weird is when a payment did go through, the invoice number I passed was included on the transaction as viewed in PayPal.

```
return actions.payment.create({
	payment: {
		transactions: [
			{
				amount: { 
					total: total, 
					currency: 'USD',
					details: {
						subtotal: amountdue,
						handling_fee: handling_fee
					}
				}
				//, invoice_number: registrationId
			}
		]
	},
	experience: {
		input_fields: {
			no_shipping: 1
		}
	}
});
```

**Comment by bluepnume** [comment]
I found the cause of the original 401 -- it's a CSRF failure with an automatic retry. I'll try to dig into the root cause of that.

Can you elaborate on the `registrationId` value and how you're ensuring it's unique? Do you generate it on page load, or during the call to `payment()` after the button is clicked?

**Comment by shicks3** [comment]
The registrationId is the id of a newly inserted database row.  submitFormRegistrationData() performs ajax call to server with all the user's form info, files it in the database, then returns that new row's id (int, currently at 1082).  I was using that to include in the PayPal calls as invoice_number so I would have a way to connect the PayPal transactions to the database.

I was trying to do it this way because I couldn't figure out how to make the connection.  PayPal transactions seem to have only a transaction id (when viewed or downloaded on the website).  What I get back from the payments call in javascript is a payment id.  Am I missing something?

In any case, docs say I can include the invoice_number in the payment package to payment.create(), only works 5% of the time.

Thanks for looking into it!

**Comment by bluepnume** [comment]
> submitFormRegistrationData() performs ajax call to server with all the user's form info, files it in the database, then returns that new row's id 

It looks like you're not waiting for that call to complete before you call `actions.payment.create()`. So there would be no chance for the `actions.payment.create()` to pass the latest `registrationId`.

Ideally, `submitFormRegistrationData` should return a promise, then you can do:

```
payment function (data, actions) {
    return submitFormRegistrationData().then(function() {
        return actions.payment.create({ ... });
    });
}
```

**Comment by shicks3** [comment]
Thank you for looking into this.   The whole problem was indeed the submitRegistrationData() function.   It didn't work to do a promise because sometimes it would take too long and cause paypal timeouts.  So I separated it into getRegistrationId() which  is called when page is built.  I reverted to calling submitRegistrationData() the old way because now I don't really care how long it takes, I've already gotten the id to submit with the paypal call.  I just have to manage orphaned ids, no biggie.

Thanks again for your help.  I learned alot!!
